torch/lib/c10d/ProcessGroupGloo.hpp

```diff
@@ -91,6 +95,27 @@ struct AlgorithmEntry {
   std::vector<at::Tensor> dst;
   std::function<void()> run;
 
+  // For CUDA tensors, the following happens:
+  //
+  // - Input tensor A is copied to persistent tensor B on stream S1
+  //   (the async stream associated with the device that stores A).
+  // - Event E1 records stream S1 so that the copy can be synchronized.
+  // - On a worker thread, stream S2 waits for event E1, executes the
+  //   CUDA-aware Gloo algorithm, and then records event E2.
+  //
+  // This approach means the caller of the process group function can
+  // retain asynchrony (no need for synchronizing its CUDA streams).
+  // Once the wait() function on the associated work object returns
+  // true, the caller can launch new CUDA kernels and they will be
+  // correctly sequenced.
+  //
+  // To do so, we need to hold on to a stream and an event per tensor.
+  // The stream is passed down to the Gloo algorithm. The event is
+  // used to record completion of the memory copy of the input tensor.
+  //
+  std::vector<CUDAStream> streams;
+  std::vector<CUDAEvent> startEvents;
```

Comment:
Are those events/streams E1/S1 or E2/S2 from the description above?

Reply:
Yes, updating the comment to be clearer.